Assign tag only allows one filter

The following code throws an exception parsing the template:

`Template template = Template.parse("{% assign v = 1 | minus: 1 | plus: 1 %}");`

The exception is:

```
MismatchedTokenException(85!=95)
        at liqp.parser.LiquidParser.reportError(LiquidParser.java:182)
        at liqp.parser.LiquidParser.assignment(LiquidParser.java:5187)
        at liqp.parser.LiquidParser.atom(LiquidParser.java:478)
        at liqp.parser.LiquidParser.block(LiquidParser.java:307)
        at liqp.parser.LiquidParser.parse(LiquidParser.java:213)
        at liqp.Template.<init>(Template.java:78)
        at liqp.Template.parse(Template.java:127)
Caused by: MismatchedTokenException(85!=95)
        at org.antlr.runtime.BaseRecognizer.recoverFromMismatchedToken(BaseRecognizer.java:617)
        at org.antlr.runtime.BaseRecognizer.match(BaseRecognizer.java:115)
        at liqp.parser.LiquidParser.assignment(LiquidParser.java:5139)
```
This template does parse with the Ruby version of Liquid.

A simple workaround is to use multiple assign statements for every filter you want to apply.

I'm not super familiar with ANTLR, but I assume the following assignment definition would fix the problem:

```
assignment
 : tagStart Assign Id EqSign expr filter? TagEnd
 ;
```

```
assignment
 : tagStart Assign Id EqSign expr filter* TagEnd
 ;
```

Thanks for the great library, the ANTLRv4 version is a great improvement too!

--- a/src/main/antlr4/liquid/parser/v4/LiquidParser.g4
+++ b/src/main/antlr4/liquid/parser/v4/LiquidParser.g4
@@ -155,7 +155,7 @@ param_expr
  ;
 
 assignment
- : tagStart Assign Id EqSign expr filter? TagEnd
+ : tagStart Assign Id EqSign expr filter* TagEnd
  ;
 
 expr

--- a/src/main/java/liqp/nodes/TagNode.java
+++ b/src/main/java/liqp/nodes/TagNode.java
@@ -6,12 +6,17 @@ import liqp.tags.Include;
 import liqp.tags.Tag;
 
 import java.io.File;
+import java.util.List;
 import java.util.Map;
 
 public class TagNode implements LNode {
 
     private Tag tag;
     private LNode[] tokens;
+
+    public TagNode(Tag tag, List<LNode> tokens) {
+        this(tag.name, tag, tokens.toArray(new LNode[tokens.size()]));
+    }
 
     public TagNode(Tag tag, LNode... tokens) {
         this(tag.name, tag, tokens);

--- a/src/main/java/liqp/parser/v4/NodeVisitor.java
+++ b/src/main/java/liqp/parser/v4/NodeVisitor.java
@@ -444,16 +444,23 @@ public class NodeVisitor extends LiquidParserBaseVisitor<LNode> {
   }
 
   // assignment
-  //  : tagStart Assign Id EqSign expr filter? TagEnd
+  //  : tagStart Assign Id EqSign expr filter* TagEnd
   //  ;
   @Override
   public LNode visitAssignment(AssignmentContext ctx) {
 
     AtomNode idNode = new AtomNode(ctx.Id().getText());
-    FilterNode filterNode = ctx.filter() == null ? null : visitFilter(ctx.filter());
     LNode exprNode = visit(ctx.expr());
+    List<LNode> allNodes = new ArrayList<>();
 
-    return new TagNode(tags.get("assign"), idNode, filterNode, exprNode);
+    allNodes.add(idNode);
+    allNodes.add(exprNode);
+
+    for (FilterContext filterContext : ctx.filter()) {
+      allNodes.add(visit(filterContext));
+    }
+
+    return new TagNode(tags.get("assign"), allNodes);
   }
 
   // expr

--- a/src/main/java/liqp/tags/Assign.java
+++ b/src/main/java/liqp/tags/Assign.java
@@ -13,21 +13,11 @@ class Assign extends Tag {
     public Object render(TemplateContext context, LNode... nodes) {
 
         String id = String.valueOf(nodes[0].render(context));
-
-        FilterNode filter = null;
-        LNode expression;
-
-        if(nodes.length >= 3) {
-            filter = (FilterNode)nodes[1];
-            expression = nodes[2];
-        }
-        else {
-            expression = nodes[1];
-        }
-
+        LNode expression = nodes[1];
         Object value = expression.render(context);
 
-        if(filter != null) {
+        for (int i = 2; i < nodes.length; i++) {
+            FilterNode filter = (FilterNode)nodes[i];
             value = filter.apply(value, context);
         }
 

--- a/src/test/java/liqp/tags/AssignTest.java
+++ b/src/test/java/liqp/tags/AssignTest.java
@@ -1,14 +1,9 @@
 package liqp.tags;
 
-import java.util.HashMap;
-import java.util.Map;
 import liqp.Template;
 import org.antlr.runtime.RecognitionException;
 import org.junit.Test;
 
-import javax.swing.table.TableModel;
-
-import static liqp.TestUtils.getNode;
 import static org.hamcrest.CoreMatchers.is;
 import static org.junit.Assert.assertThat;
 

--- a/src/test/java/liqp/tags/AssignTest.java
+++ b/src/test/java/liqp/tags/AssignTest.java
@@ -6,6 +6,8 @@ import liqp.Template;
 import org.antlr.runtime.RecognitionException;
 import org.junit.Test;
 
+import javax.swing.table.TableModel;
+
 import static liqp.TestUtils.getNode;
 import static org.hamcrest.CoreMatchers.is;
 import static org.junit.Assert.assertThat;
@@ -63,6 +65,12 @@ public class AssignTest {
         assertThat(Template.parse("{% assign foo = values %}.{{ foo[1] }}.").render("values", values), is(".bar."));
 
         assertThat(Template.parse("{% assign foo = values | split: \",\" %}.{{ foo[1] }}.").render("values", "foo,bar,baz"), is(".bar."));
+    }
+
+    @Test
+    public void multipleFiltersTest() {
+        // https://github.com/bkiers/Liqp/issues/84
+        assertThat(Template.parse("{% assign v = 1 | minus: 10 | plus: 5 %}{{v}}").render(), is("-4"));
     }
 
     /*

